src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h

```diff
@@ -78,19 +76,69 @@ class HybridSchedulingPolicy : public ISchedulingPolicy {
+  bool IsNodeFeasible(const scheduling::NodeID &node_id,
+                      const NodeFilter &node_filter,
+                      const NodeResources &node_resources,
+                      const ResourceRequest &resource_request) const;
+
+  /// helper function compute a score between 0-1 indicates
+  /// the preference of the current node (the lower score,
+  /// the more preferable.
+  float CompuateLocalNodeScore(float spread_threshold) const;
+
+  scheduling::NodeID GetBestNode(
+      std::vector<std::pair<scheduling::NodeID, float>> &node_scores,
+      size_t num_candidate_nodes,
+      bool prioritize_local_node,
+      float local_node_score) const;
+
   /// \param resource_request: The resource request we're attempting to schedule.
+  /// \param spread_threshold: The fraction of resource utilization on a node after
+  /// which the scheduler starts to prefer spreading tasks to other nodes.
+  //// This balances between locality and
+  /// even balancing of load. Low values (min 0.0) encourage more load spreading.
+  /// \param force_spillback: don't schedule on local node if true.
+  /// \param require_available: If true, only schedule on nodes who have resources
+  /// available to fulfill the request. Otherwise, schedule on nodes whose resources
+  /// capacity can fulfill the request, even if the resources are not currently
+  /// available.
   /// \param node_filter: defines the subset of nodes were are allowed to schedule on.
   /// can be one of kAny (can schedule on all nodes), kGPU (can only schedule on kGPU
   /// nodes), kNonGpu (can only schedule on non-GPU nodes.
+  /// \param schedule_top_k_absolute: scheduler will randomly pick
+  /// one node from the top k in the cluster to improve load balancing. The
+  /// scheduler guarantees k is at least equal to schedule_top_k_absolute.
+  /// \param schedule_top_k_fraction: The scheduler will randomly pick
+  /// one node from the top k in the cluster to improve load balancing. The
+  /// scheduler guarantees k is at least equal to this fraction * the number of
+  /// nodes in the cluster.
   ///
   /// \return -1 if the task is unfeasible, otherwise the node id (key in `nodes`) to
   /// schedule on.
-  scheduling::NodeID HybridPolicyWithFilter(const ResourceRequest &resource_request,
-                                            float spread_threshold,
-                                            bool force_spillback,
-                                            bool require_available,
-                                            const std::string &preferred_node,
-                                            NodeFilter node_filter = NodeFilter::kAny);
+  scheduling::NodeID ScheduleImpl(const ResourceRequest &resource_request,
+                                  float spread_threshold,
+                                  bool force_spillback,
+                                  bool require_available,
+                                  NodeFilter node_filter,
```

Comment:
same q here. Removing preferred node id means we removed the contribution from Ant?